src/Processors/Formats/Impl/MsgPackRowInputFormat.cpp

```diff
@@ -328,21 +375,47 @@ bool MsgPackVisitor::visit_boolean(bool value)
 
 bool MsgPackVisitor::start_array(size_t size) // NOLINT
 {
-    if (!isArray(info_stack.top().type))
-        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Cannot insert MessagePack array into column with type {}.", info_stack.top().type->getName());
+    if (isArray(info_stack.top().type))
+    {
+        auto nested_type = assert_cast<const DataTypeArray &>(*info_stack.top().type).getNestedType();
+        ColumnArray & column_array = assert_cast<ColumnArray &>(info_stack.top().column);
+        ColumnArray::Offsets & offsets = column_array.getOffsets();
+        IColumn & nested_column = column_array.getData();
+        offsets.push_back(offsets.back() + size);
+        if (size > 0)
+            info_stack.push(Info{nested_column, nested_type, false, size, nullptr});
+    }
+    else if (isTuple(info_stack.top().type))
+    {
+        const auto & tuple_type = assert_cast<const DataTypeTuple &>(*info_stack.top().type);
+        const auto & nested_types = tuple_type.getElements();
+        if (size != nested_types.size())
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Cannot insert MessagePack array with size {} into Tuple column with {} elements", size, nested_types.size());
+
+        ColumnTuple & column_tuple = assert_cast<ColumnTuple &>(info_stack.top().column);
+        /// Push nested columns into stack in reverse order.
+        for (ssize_t i = nested_types.size() - 1; i >= 0; --i)
+            info_stack.push(Info{column_tuple.getColumn(i), nested_types[i], true, std::nullopt, nullptr});
+    }
+    else
+    {
+        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Cannot insert MessagePack array into column with type {}", info_stack.top().type->getName());
+    }
 
-    auto nested_type = assert_cast<const DataTypeArray &>(*info_stack.top().type).getNestedType();
-    ColumnArray & column_array = assert_cast<ColumnArray &>(info_stack.top().column);
-    ColumnArray::Offsets & offsets = column_array.getOffsets();
-    IColumn & nested_column = column_array.getData();
-    offsets.push_back(offsets.back() + size);
-    info_stack.push(Info{nested_column, nested_type, nullptr});
     return true;
 }
 
-bool MsgPackVisitor::end_array() // NOLINT
+
+bool MsgPackVisitor::end_array_item() // NOLINT
 {
-    info_stack.pop();
+    if (info_stack.top().is_tuple_element)
+        info_stack.pop();
+    else
+    {
+        --(*info_stack.top().array_size);
+        if (*info_stack.top().array_size == 0)
```

Comment:
```suggestion
        assert(info_stack.top().array_size.has_value());
        auto & current_array_size = *info_stack.top().array_size;
        --current_array_size;
        if (current_array_size == 0)
```

maybe something like this?
assert to be more clear about the assumption and